### PR TITLE
C-API request deserialization

### PR DESCRIPTION
**Pull request by atobiszei**
TODO:
* binary inputs

JIRA:CVS-95304

### CODE REVIEW

File: src/deserialization.hpp [hunk 1]
```diff
@@ -404,4 +446,55 @@ Status deserializePredictRequest(
...
+template <class TensorProtoDeserializator, class Sink>
+Status deserializePredictRequest(
+    const InferenceRequest& request,
+    const tensor_map_t& inputMap,
+    Sink& inputSink, bool isPipeline) {
+    OVMS_PROFILE_FUNCTION();
+    Status status;
+    for (const auto& pair : inputMap) {
+        try {
+            const auto& name = pair.first;
```
Comment (dkalinowski) [inline src/deserialization.hpp:458]: You can use cpp17 feature `for (const auto& [name, tensorInfo] : inputMap) {` to save 2 lines of code

File: src/deserialization.hpp [hunk 2]
```diff
@@ -404,4 +446,55 @@ Status deserializePredictRequest(
...
+            auto tensorInfo = pair.second;
+            const InferenceTensor* requestInputPtr{nullptr};
+            auto status = request.getInput(name.c_str(), &requestInputPtr);
```
Comment (dkalinowski) [inline src/deserialization.hpp:461]: You do not check status after getInput call

File: src/pocapiinternal.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#pragma once
```
Comment (dkalinowski) [inline src/pocapiinternal.hpp:1]: At some point we started to use pragme before license header

File: src/test/deserialization_tests.cpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 //*****************************************************************************
-// Copyright 2020 Intel Corporation
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:2]: Here you edit license year but not in deserialization.cpp/hpp

File: src/test/deserialization_tests.cpp [hunk 5]
```diff
@@ -74,7 +77,7 @@ class TensorflowGRPCPredict : public ::testing::TestWithParam<ovms::Precision> {
         tensorShape->Clear();
         tensorShape->add_dim()->set_size(1);
         tensorShape->add_dim()->set_size(DUMMY_MODEL_INPUT_SIZE);
-        *(tensorProto.mutable_tensor_content()) = std::string(1 * DUMMY_MODEL_INPUT_SIZE, '1');
+        *(tensorProto.mutable_tensor_content()) = std::string(1 * sizeof(float) * DUMMY_MODEL_INPUT_SIZE, '1');
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:80]: How did it work before with smaller buffer?

File: src/test/deserialization_tests.cpp [hunk 6]
```diff
@@ -85,6 +88,70 @@ class TensorflowGRPCPredict : public ::testing::TestWithParam<ovms::Precision> {
...
+    void SetUp() override {
+        auto precision = ovms::Precision::FP32;
+        tensorMap[tensorName] = std::make_shared<ovms::TensorInfo>(
+            tensorName,
+            precision,
+            shape_t{1, 3},
+            Layout{"NC"});
+        SetUpTensorProto(getPrecisionAsOVMSDataType(precision));
+    }
+    void SetUpTensorProto(DataType dataType) {
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:102]: You have hardcoded FP32 here, and you do not use precision param

File: src/test/deserialization_tests.cpp [hunk 7]
```diff
@@ -85,6 +88,70 @@ class TensorflowGRPCPredict : public ::testing::TestWithParam<ovms::Precision> {
...
 class DeserializeTFTensorProtoNegative : public TensorflowGRPCPredict {};
 
+class CAPIPredict : public ::testing::TestWithParam<ovms::Precision> {
+protected:
+    void SetUp() override {
+        auto precision = ovms::Precision::FP32;
+        tensorMap[tensorName] = std::make_shared<ovms::TensorInfo>(
+            tensorName,
+            precision,
+            shape_t{1, 3},
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:98]: Shape mismatch - tensor map contains shape [1, 3] but in line 103 you declare [1, 10]. Why does the test pass?

File: src/test/deserialization_tests.cpp [hunk 8]
```diff
@@ -85,6 +88,70 @@ class TensorflowGRPCPredict : public ::testing::TestWithParam<ovms::Precision> {
...
+const std::string CAPIPredictRequest::DATA{1 * sizeof(float) * DUMMY_MODEL_INPUT_SIZE, '1'};
+// dispose connection to CAPI Predict
+TEST_F(CAPIPredictRequest, ShouldSuccessForSupportedPrecision) {
+    ov::Core ieCore;
+    std::shared_ptr<ov::Model> model = ieCore.read_model(std::filesystem::current_path().u8string() + "/src/test/dummy/1/dummy.xml");
+    ov::CompiledModel compiledModel = ieCore.compile_model(model, "CPU");
+    ov::InferRequest inferRequest = compiledModel.create_infer_request();
+    InputSink<ov::InferRequest&> inputSink(inferRequest);
+    auto status = deserializePredictRequest<ConcreteTensorProtoDeserializator>(request, tensorMap, inputSink, isPipeline);
+    EXPECT_TRUE(status.ok());
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:149]: Is ok() check enough to verify if precision deserialization works? You can set wrong precision and still the test will pass

File: src/test/deserialization_tests.cpp [hunk 9]
```diff
@@ -193,6 +260,15 @@ TEST_P(DeserializeTFTensorProtoNegative, ShouldReturnNullptrForPrecision) {
                                << " should return nullptr";
 }
 
+TEST_P(DeserializeCAPITensorProtoNegative, ShouldReturnNullptrForPrecision) {
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:263]: The test is NEgative but named to be positive: `ShouldReturnNullptrForPrecision`

File: src/test/deserialization_tests.cpp [hunk 10]
```diff
@@ -472,6 +557,14 @@ INSTANTIATE_TEST_SUITE_P(
         return toString(info.param);
     });
 
+INSTANTIATE_TEST_SUITE_P(
+    TestDeserialize,
+    CAPIPredictRequest,
+    ::testing::ValuesIn(SUPPORTED_KFS_INPUT_PRECISIONS),
```
Comment (dkalinowski) [inline src/test/deserialization_tests.cpp:563]: This is CAPI test but you use KFS precision vector

File: src/deserialization.hpp [hunk 11]
```diff
@@ -210,6 +215,36 @@ class ConcreteTensorProtoDeserializator {
...
+        case ovms::Precision::FP16:
+        case ovms::Precision::I64:
+        case ovms::Precision::I32:
+        case ovms::Precision::I16:
+        case ovms::Precision::I8:
+        case ovms::Precision::U64:
+        case ovms::Precision::U32:
+        case ovms::Precision::U16:
+        case ovms::Precision::BOOL:
+        case ovms::Precision::U1:
```
Comment (dkalinowski) [inline src/deserialization.hpp:234]: Do we test U1 / BOOL precision? I have added those precisions to unsupported in serialization.

File: src/deserialization.hpp [hunk 12]
```diff
@@ -30,6 +30,8 @@
 #pragma GCC diagnostic pop
 
 #include "binaryutils.hpp"
+#include "inferencerequest.hpp"
+#include "inferencetensor.hpp"
 #include "profiler.hpp"
```
Comment (dkalinowski) [inline src/deserialization.hpp:35]: Missing license year change as well. Please look up remaining files

**Review by rasapala — APPROVED** [review]
LGTM

File: src/pocapi.hpp [hunk 13]
```diff
@@ -46,7 +46,8 @@ enum DataType {
     OVMS_DATATYPE_DYNAMIC,
     OVMS_DATATYPE_MIXED,
     OVMS_DATATYPE_Q78,
-    OVMS_DATATYPE_BIN
+    OVMS_DATATYPE_BIN,
+    OVMS_DATATYPE_END
```
Comment (dkalinowski) [inline src/pocapi.hpp:50]: Are we sure we want to have END in exposed api header? It blocks us from adding more precisions later

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**